### PR TITLE
Wireguard remote mechanism for kernel forwarder

### PR DESCRIPTION
**Pull request by artem-belov**
## Motivation and Context
#2069

## How Has This Been Tested?

- [ ] Covered by existing integration testing
- [x] Added integration testing to cover
- [ ] Tested locally
- [ ] Have not tested

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: forwarder/kernel-forwarder/pkg/kernelforwarder/local/local.go [hunk 1]
```diff
@@ -0,0 +1,72 @@
+// Copyright (c) 2020 Doc.ai and/or its affiliates.
```
Comment (nickolaev) [inline forwarder/kernel-forwarder/pkg/kernelforwarder/local/local.go]: Please keep the license of the code when you move it to new files. Having two Copyrights is completely fine in such a case when you introduce significant changes too.

File: forwarder/kernel-forwarder/pkg/kernelforwarder/remote/remote.go [hunk 2]
```diff
@@ -0,0 +1,65 @@
+// Copyright (c) 2020 Doc.ai and/or its affiliates.
```
Comment (nickolaev) [inline forwarder/kernel-forwarder/pkg/kernelforwarder/remote/remote.go]: Please keep the license of the code if you move it around.

File: forwarder/kernel-forwarder/pkg/kernelforwarder/remote/vxlan.go [hunk 3]
```diff
@@ -0,0 +1,79 @@
+// Copyright (c) 2020 Doc.ai and/or its affiliates.
```
Comment (nickolaev) [inline forwarder/kernel-forwarder/pkg/kernelforwarder/remote/vxlan.go]: Please keep the license of the code if you move it around.

File: forwarder/kernel-forwarder/pkg/kernelforwarder/local/local.go (hunk 1 above)
Comment (artem-belov) [reply forwarder/kernel-forwarder/pkg/kernelforwarder/local/local.go]: Done

File: forwarder/kernel-forwarder/pkg/kernelforwarder/remote/remote.go (hunk 2 above)
Comment (artem-belov) [reply forwarder/kernel-forwarder/pkg/kernelforwarder/remote/remote.go]: Done

File: forwarder/kernel-forwarder/pkg/kernelforwarder/remote/vxlan.go (hunk 3 above)
Comment (artem-belov) [reply forwarder/kernel-forwarder/pkg/kernelforwarder/remote/vxlan.go]: Done

File: forwarder/kernel-forwarder/pkg/kernelforwarder/kernelforwarder.go [hunk 4]
```diff
@@ -59,7 +66,13 @@ func (k *KernelForwarder) CreateForwarderServer(config *common.ForwarderConfig)
...
 func (k *KernelForwarder) Request(ctx context.Context, crossConnect *crossconnect.CrossConnect) (*crossconnect.CrossConnect, error) {
 	logrus.Infof("Request() called with %v", crossConnect)
-	err := k.connectOrDisconnect(crossConnect, cCONNECT)
+
+	if err := crossConnect.IsValid(); err != nil {
+		logrus.Errorf("Close: %v is not valid, reason: %v", crossConnect, err)
+		return crossConnect, err
+	}
+
+	err := k.connectOrDisconnect(crossConnect, true)
```
Comment (rdimitrov) [inline forwarder/kernel-forwarder/pkg/kernelforwarder/kernelforwarder.go]: @artem-belov 
Is it possible to make it more explicit here? 
I'm open to use something else and ditch the connect/disconnect labels as long as it's better than true/false 👍

File: forwarder/kernel-forwarder/pkg/kernelforwarder/remote/wireguard.go [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+	remotePublicKey, err := wgtypes.ParseKey(remotePublicKeyStr)
+	if err != nil {
+		return errors.Errorf("failed to parse remote public key: %v", err)
+	}
+
+	wgDevice, err := createWireguardDevice(ifaceName)
+	if err != nil {
+		return errors.Errorf("Wireguard error: %v", err)
+	}
+	//defer wgDevice.Close()
```
Comment (denis-tingajkin) [inline forwarder/kernel-forwarder/pkg/kernelforwarder/remote/wireguard.go]: TODO: remove commented code

File: forwarder/kernel-forwarder/pkg/kernelforwarder/remote/wireguard.go [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+	if err != nil {
+		wgDevice.Close()
+		return errors.Errorf("Wireguard error: %v", err)
+	}
+
+	return nil
+}
+
+func (c *Connect) deleteWireguardInterface(ifaceName string) error {
+	if wgDevice, ok := c.wireguardDevices[ifaceName]; ok {
```
Comment (denis-tingajkin) [inline forwarder/kernel-forwarder/pkg/kernelforwarder/remote/wireguard.go]: TODO: fix: No one not writing in this map

File: forwarder/kernel-forwarder/pkg/kernelforwarder/kernelforwarder.go (hunk 4 above)
Comment (denis-tingajkin) [reply forwarder/kernel-forwarder/pkg/kernelforwarder/kernelforwarder.go]: I also think we can solve this problem on kernel-sdk level.

**Comment by denis-tingajkin** [comment]
Here failed one test `TestVPNNSERemote` it is a known issue https://github.com/networkservicemesh/networkservicemesh/issues/2029

File: .circleci/config.yml [hunk 7]
```diff
@@ -102,6 +102,7 @@ jobs:
 
 # integration
   integration-tests:
+    resource_class: xlarge
```
Comment (edwarnicke) [inline .circleci/config.yml]: Did we fail without the larger class size here? Why the increase resource usage?

**Pull request merged**